### PR TITLE
WIP update order model

### PR DESCRIPTION
**Pull request by tasawernawaz**
ECOM-2236
- Update the model `Order` and the order number generation class `OrderNumberGenerator` logic. 
- Add migration for the change in the model `Order`
- Update related tests.

This PR is dependent on (ECOM-2232): https://github.com/edx/ecommerce/pull/326
Docs: https://github.com/edx/ecommerce/pull/341

### CODE REVIEW

File: ecommerce/extensions/order/models.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from django.db import models
 from django.utils.translation import ugettext_lazy as _
 from oscar.apps.order.abstract_models import AbstractOrder, AbstractPaymentEvent, AbstractLine
+from oscar.apps.partner.models import Partner
```
Comment (zubair-arbi) [inline ecommerce/extensions/order/models.py]: Please order the imports

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def add_partner_and_update_order_number(apps, schema_editor):
+    """ Updating existing orders. Adding relation between order and partner.
+    """
+
+    Order = apps.get_model('order', 'Order')
+    Partner = apps.get_model('partner', 'Partner')
+
+    partner, __ = Partner.objects.get_or_create(code=PARTNER_CODE)
```
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: You should NOT be creating a partner. All you need to do here is update `Order.partner` for existing orders.

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    orders = Order.objects.all()
+
+    # Update all existing basket with edx partner
+    orders.update(partner=partner)
```
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: @feanil do we have a way to test a query that updates all 1M+ order rows?

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    # Update all orders to add PARTNER_CODE as a prefix to order numbers
+    for order in orders:
```
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: The order numbers already have a prefix.

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        order.number = '{partner_code}-{order_number}'.format(partner_code=PARTNER_CODE,
+                                                              order_number=order.number.rsplit('-')[-1])
+        order.save()
+
+
+def remove_partner_and_update_order_number(apps, schema_editor):
+    # This method is for backward compatibility, It will remove partner reference_id
+    # and add ORDER_NUMBER_PREFIX to the order numbers
+
+    Order = apps.get_model('order', 'Order')
```
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: This should be a no-op.

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='historicalorder',
+            name='partner',
+            field=models.ForeignKey(related_name='+', on_delete=django.db.models.deletion.DO_NOTHING, db_constraint=False, blank=True, to='partner.Partner', null=True),
+        ),
+        migrations.AddField(
+            model_name='order',
+            name='partner',
+            field=models.ForeignKey(default=1, to='partner.Partner'),
```
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: This is not a valid default. There should NOT be a default value.

File: ecommerce/extensions/order/models.py [hunk 7]
```diff
@@ -1,14 +1,18 @@
 # noinspection PyUnresolvedReferences
 from django.db import models
 from django.utils.translation import ugettext_lazy as _
-from oscar.apps.order.abstract_models import AbstractOrder, AbstractPaymentEvent, AbstractLine
```
Comment (clintonb) [inline ecommerce/extensions/order/models.py:4]: Revert this change. Imports should be should be grouped (native, third-party, project) and alphabetized.

File: ecommerce/extensions/order/models.py [hunk 8]
```diff
@@ -1,14 +1,18 @@
...
+
 from simple_history.models import HistoricalRecords
 
+from oscar.apps.order.abstract_models import AbstractOrder, AbstractPaymentEvent, AbstractLine
+from oscar.apps.partner.models import Partner
```
Comment (clintonb) [inline ecommerce/extensions/order/models.py]: Do not import models in this manner. See https://django-oscar.readthedocs.org/en/releases-1.1/topics/class_loading_explained.html?highlight=get_class.

File: ecommerce/extensions/order/models.py [hunk 9]
```diff
@@ -1,14 +1,18 @@
...
+
 from ecommerce.extensions.fulfillment.status import ORDER
 
 
 class Order(AbstractOrder):
     history = HistoricalRecords()
+    partner = models.ForeignKey(Partner)
```
Comment (clintonb) [inline ecommerce/extensions/order/models.py]: Use `partner.Partner`.

File: ecommerce/extensions/order/models.py (hunk 7 above)
Comment (zubair-arbi) [inline ecommerce/extensions/order/models.py:4]: @clintonb the import order was changed as suggested by this guideline: https://github.com/edx/edx-platform/wiki/Python-Guidelines#imports-order

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py (hunk 4 above)
Comment (tasawernawaz) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: @clintonb Here we need to add partner code as prefix instead of our previous prefix 'OSCAR'.
Similarly for backward compatibility I am adding again 'OSCAR' and removing partner code.

**Comment by clintonb** [comment]
ECOM-2232 (basket model updates) must be finished before work can be completed on the order model.

File: ecommerce/extensions/order/models.py (hunk 7 above)
Comment (clintonb) [inline ecommerce/extensions/order/models.py:4]: That's a rather strange interpretation of PEP8. We aren't that strict. Django and Oscar are both third-party libraries and should be grouped together. I'll bring this up in a later architecture meeting.

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py (hunk 4 above)
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: We don't use `OSCAR`. We have an override for that setting in stage and production, where it is set to `EDX`.

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py (hunk 2 above)
Comment (tasawernawaz) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: I am using here `get_or_create` method, because there must be at least one partner that will be referenced for existing orders.
Do we assume that there will be partner in Partner modal ? In that case we simple update orders.

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py (hunk 5 above)
Comment (tasawernawaz) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: @clintonb For backward compatibility I am adding 'OSCAR' and removing partner code. Do you think we should keep existing order numbers without updating them for partner code as prefix. In that case we can simply pass this method.

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py (hunk 4 above)
Comment (tasawernawaz) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: @clintonb Thanks. So we can remove forward and backward data migration for `order.numbers`.

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py (hunk 2 above)
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: We already have an edX partner. Use `Partner.objects.get(id=1)`.

File: ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py (hunk 3 above)
Comment (feanil) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: Loadtest environment with a loadtest running against the ecommerce app.

Comment (clintonb) [inline ecommerce/extensions/order/migrations/0010_auto_20150911_0905.py]: Thanks @feanil. @tasawernawaz let me know if you need help with this.

**Comment by zubair-arbi** [comment]
@tasawernawaz change looks good. But we will need to wait for the PR: https://github.com/edx/ecommerce/pull/326 to merge first.

**Comment by clintonb** [comment]
@tasawernawaz remember to squash your commits.

File: ecommerce/extensions/order/migrations/0011_auto_20150915_0336.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('order', '0010_auto_20150915_0332'),
+    ]
+
+    operations = [
+        migrations.AlterField(
```
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0011_auto_20150915_0336.py]: Merge this with the migration above.

File: ecommerce/extensions/order/models.py [hunk 11]
```diff
@@ -9,6 +9,7 @@
 
 class Order(AbstractOrder):
     history = HistoricalRecords()
+    partner = models.ForeignKey('partner.Partner', related_name='partner')
```
Comment (clintonb) [inline ecommerce/extensions/order/models.py]: related_name should be `'orders'`.

File: ecommerce/extensions/order/tests/test_utils.py [hunk 12]
```diff
@@ -22,8 +19,11 @@ def test_order_number_generation(self):
         first_order_number = OrderNumberGenerator().order_number(first_basket)
         second_order_number = OrderNumberGenerator().order_number(second_basket)
 
-        self.assertIn(self.ORDER_NUMBER_PREFIX, first_order_number)
-        self.assertIn(self.ORDER_NUMBER_PREFIX, second_order_number)
+        # TODO add partner code instead of hard coded value
```
Comment (clintonb) [inline ecommerce/extensions/order/tests/test_utils.py:22]: Do not merge until this is fixed.

File: ecommerce/extensions/order/utils.py [hunk 13]
```diff
@@ -38,7 +46,10 @@ def basket_id(self, order_number):
         Returns:
             int: The basket ID used to generate the provided order number.
         """
-        order_id = int(order_number.lstrip(u'{prefix}-'.format(prefix=settings.ORDER_NUMBER_PREFIX)))
+
+        # here we are removing partner code from order number and converting back into integer
+        # for getting basket_id, e.g order_number = EDX-100001
+        order_id = int(order_number.rsplit('-')[-1])
```
Comment (clintonb) [inline ecommerce/extensions/order/utils.py]: This will be slightly more legible if you use `[1]` as your index instead of `[-1]`.

File: ecommerce/extensions/order/utils.py [hunk 14]
```diff
@@ -38,7 +46,10 @@ def basket_id(self, order_number):
         Returns:
             int: The basket ID used to generate the provided order number.
         """
-        order_id = int(order_number.lstrip(u'{prefix}-'.format(prefix=settings.ORDER_NUMBER_PREFIX)))
+
+        # here we are removing partner code from order number and converting back into integer
```
Comment (clintonb) [inline ecommerce/extensions/order/utils.py]: The docstring already explains what's going on in this method. The extra comment is really necessary.

File: ecommerce/settings/local.py [hunk 15]
```diff
@@ -12,9 +12,6 @@
 # DEBUG CONFIGURATION
 # See: https://docs.djangoproject.com/en/dev/ref/settings/#debug
 DEBUG = True
-
-# See: https://docs.djangoproject.com/en/dev/ref/settings/#template-debug
-TEMPLATE_DEBUG = DEBUG
```
Comment (clintonb) [inline ecommerce/settings/local.py]: Make sure you've rebased on master. This setting was removed sometime last week.

File: ecommerce/extensions/order/utils.py (hunk 13 above)
Comment (tasawernawaz) [inline ecommerce/extensions/order/utils.py]: I have used this because a partner code may contain '-', in that we case may get unexpected result.

Comment (clintonb) [inline ecommerce/extensions/order/utils.py]: The partner's code should only be alphanumeric characters.

Comment (tasawernawaz) [inline ecommerce/extensions/order/utils.py]: @clintonb I have confirmed this from #325. We have partner.short_code as Char field.

File: ecommerce/extensions/order/migrations/0010_auto_20150915_0332.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='historicalorder',
+            name='partner',
+            field=models.ForeignKey(related_name='+', on_delete=django.db.models.deletion.DO_NOTHING, db_constraint=False, blank=True, to='partner.Partner', null=True),
+        ),
+        migrations.AddField(
+            model_name='order',
+            name='partner',
+            field=models.ForeignKey(related_name='orders', to='partner.Partner', null=True, blank=True),
```
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0010_auto_20150915_0332.py]: I was reminded that you can do all of the migrations in a single step. Delete this migration and re-run `makemigrations`. Specify 1 as the default.
